client/auth/oauth-utils.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Internal dependencies
+ */
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import { bumpStat } from 'calypso/lib/analytics/mc';
+
+export const actions = {
+	AUTH_LOGIN: 'AUTH_LOGIN',
+	RECEIVE_AUTH_LOGIN: 'RECEIVE_AUTH_LOGIN',
+	AUTH_RESET: 'AUTH_RESET',
+};
+
+export const errors = {
+	ERROR_REQUIRES_2FA: 'needs_2fa', // From WP.com API
+	ERROR_INVALID_OTP: 'invalid_otp', // From WP.com API
+};
+
+export async function makeAuthRequest( username, password, authCode = '' ) {
+	const response = await globalThis.fetch( '/oauth', {
+		method: 'POST',
+		headers: { 'Content-Type': 'application/json' },
+		body: JSON.stringify( {
+			username,
+			password,
+			auth_code: authCode.replace( /\s/g, '' ),
+		} ),
+	} );
+
+	const json = await response.json();
+	const errorMessage = json?.error_description ?? '';
+
+	if ( errorMessage ) {
+		const error = new Error( errorMessage );
+		error.type = json.error;
+
+		throw error;
+	}
```

Comment:
The original code looks at `error_description` and throws the error (or more precisely, calls the node-style callback with `error`) only if `! response.ok`. That means that the response had non-200 HTTP state.

Reply:
should be addressed in 317bfc4